Sources/PostgresNIO/New/PSQLRowSequence.swift

```diff
@@ -0,0 +1,607 @@
+            continuation.resume(throwing: error)
+            
+        case .none:
+            break
+        }
+    }
+    
+    func sequenceDeinitialized() {
+        let action = self.lock.withLock {
+            self.state.sequenceDeinitialized()
+        }
+        
+        switch action {
+        case .cancelStream(let source):
+            source.cancel()
+        case .none:
+            break
+        }
+    }
+    
+    func makeAsyncIterator() -> PSQLRowSequence.Iterator {
+        self.lock.withLock {
+            self.state.createAsyncIterator()
+        }
+        let iterator = PSQLRowSequence.Iterator(consumer: self)
+        return iterator
+    }
+
+    func iteratorDeinitialized() {
+        let action = self.lock.withLock {
+            self.state.iteratorDeinitialized()
+        }
+
+        switch action {
+        case .cancelStream(let source):
+            source.cancel()
+        case .none:
+            break
+        }
+    }
+    
+    func next() async throws -> PSQLRow? {
+        self.lock.lock()
+        switch self.state.next() {
+        case .returnNil:
+            self.lock.unlock()
+            return nil
+            
+        case .returnRow(let data, signalDemandTo: let source):
+            self.lock.unlock()
+            source?.demand()
+            return PSQLRow(
+                data: data,
+                lookupTable: self.lookupTable,
+                columns: self.columns,
+                jsonDecoder: self.jsonDecoder
+            )
+            
+        case .throwError(let error):
+            self.lock.unlock()
+            throw error
+            
+        case .hitSlowPath:
+            return try await withCheckedThrowingContinuation { continuation in
+                let slowPathAction = self.state.next(for: continuation)
+                self.lock.unlock()
+                switch slowPathAction {
+                case .signalDemand(let source):
+                    source.demand()
+                case .none:
+                    break
+                }
+            }
+        }
+    }
+    
+}
+
+extension AsyncStreamConsumer {
+    struct StateMachine {
+        enum UpstreamState {
+            enum DemandState {
+                case canAskForMore
+                case waitingForMore(CheckedContinuation<PSQLRow?, Error>?)
+            }
+            
+            case initialized
+            case streaming(AdaptiveRowBuffer, PSQLRowStream, DemandState)
+            /// The upstream has finished, but the downstream has not consumed all events.
+            case finished(AdaptiveRowBuffer, String)
+            /// The upstream has failed, but the downstream has not consumed the error yet.
+            case failed(Error)
+            /// The upstream has failed or finished and the downstream has consumed all events. Final state.
+            case consumed
+
+            case modifying
+        }
+        
+        enum DownstreamState {
```

Comment:
```suggestion
        enum DownstreamState: Hashable {
```
(for use further down)